lstm/bmi_lstm.py

```diff
@@ -1,1046 +1,562 @@
+                outputs[output.name].append(output.value[0])
+
+        # ensemble output and set output variables
+        for name, values in outputs.items():
+            self._outputs.set_value(name, np.mean(values, dtype="float64"))
+
+        # increment model timestep
+        self._timestep += 1
+
+    def update_until(self, time: float) -> None:
+        if time <= self.get_current_time():
+            current_time = self.get_current_time()
+            logger.warning(f"no update performed: {time=} <= {current_time=}")
+            return None
+
+        n_steps, remainder = divmod(
+            time - self.get_current_time(), self.get_time_step()
+        )
+
+        if remainder != 0:
+            logger.warning(
+                f"time is not multiple of time step size. updating until: {time - remainder=} "
+            )
 
-        Parameters
-        ----------
-        item
-            The name of the attribute item to get.
+        for _ in range(int(n_steps)):
+            self.update()
 
-        Returns
-        -------
-        The value of the named item.
-        """
-        # Have these work explicitly (or else loops)
-        if item == '_input_var_names' or item == '_output_var_names':
-            return super(bmi_LSTM, self).__getattribute__(item)
-
-        # By default, for things other than BMI variables, use normal behavior
-        if item not in super(bmi_LSTM, self).__getattribute__('_input_var_names') and item not in super(bmi_LSTM, self).__getattribute__('_output_var_names'):
-            return super(bmi_LSTM, self).__getattribute__(item)
-
-        # Return the single scalar value from any ndarray of size 1
-        value = super(bmi_LSTM, self).__getattribute__(item)
-        if isinstance(value, np.ndarray) and value.size == 1:
-            return value[0]
-        else:
-            return value
-
-    def __setattr__(self, key, value):
-        """
-        Customized instance attribute mutator functionality.
+    def finalize(self) -> None: ...
 
-        For those attribute with keys indicating they are a BMI input or output variable (which should be in numpy
-        arrays), wrap any scalar ``value`` as a one-element numpy array and use that in a nested call to the superclass
-        implementation of this function.  In any other cases, just pass the given ``key`` and ``value`` to a nested
-        call.
+    def get_component_name(self) -> str:
+        return "LSTM"
 
-        This supports automatically having a BMI variable be backed by a numpy array, even if it is initialized using a
-        scalar, while otherwise maintaining standard behavior.
+    def get_input_item_count(self) -> int:
+        return len(self._dynamic_inputs)
 
-        Parameters
-        ----------
-        key
-        value
+    def get_output_item_count(self) -> int:
+        return len(self._outputs)
 
-        Returns
-        -------
+    def get_input_var_names(self) -> tuple[str, ...]:  # type: ignore
```

Comment:
I may need to double check the pybind/ngen `get_input_var_names` proxy and see how flexibe it is with this return type (list vs tuple).

Reply:
FWIW its a `tuple[str, ...]` in [`bmipy`](https://github.com/csdms/bmi-python/blob/d0b9b281f4cf9948dabef3a48346a0b8ee1a352f/src/bmipy/bmi.py#L111).

Reply:
Double checked the model engine's call of this, and it indeed types this as a `py::tuple`
https://github.com/NOAA-OWP/ngen/blob/4fb5bb68dc397298bca470dfec94db2c1dcb42fe/src/bmi/Bmi_Py_Adapter.cpp#L59